src/ILLink.Shared/DataFlow/ValueSet.cs

```diff
@@ -9,49 +10,151 @@ namespace ILLink.Shared.DataFlow
 	public readonly struct ValueSet<TValue> : IEquatable<ValueSet<TValue>>, IEnumerable<TValue>
 		where TValue : notnull
 	{
-		public readonly HashSet<TValue>? Values;
+		// Since we're going to do lot of type checks for this class a lot, it is much more efficient
+		// if the class is sealed (as then the runtime can do a simple method table pointer comparison)
+		class SealedHashSet : HashSet<TValue>
```

Comment:
Why not call it EnumerableValues to make it clear what it does?
I'd also seal the class